ntoskrnl/io/iomgr/deviface.c

```diff
@@ -28,6 +29,257 @@ IopGetDeviceObjectFromDeviceInstance(PUNICODE_STRING DeviceInstance);
+    UNICODE_STRING SymbolicLink;
+    NTSTATUS Status;
+
+    /* Use a backslash if reference string is not specified */
+    if (!ReferenceString)
+        ReferenceString = &PathSep;
+
+    /* Build up new symbolic link */
+    SymbolicLink.Length = 0;
+    SymbolicLink.MaximumLength =
+        PrefixString->Length + DeviceString->Length
+        + sizeof(L"#") + GuidString->Length
+        + ReferenceString->Length
+        + sizeof(UNICODE_NULL);
+    SymbolicLink.Buffer = ExAllocatePoolWithTag(PagedPool, SymbolicLink.MaximumLength, TAG_IO);
+    if (!SymbolicLink.Buffer)
+    {
+        DPRINT1("ExAllocatePoolWithTag() failed\n");
+        return STATUS_INSUFFICIENT_RESOURCES;
+    }
+
+    Status = RtlUnicodeStringPrintf(&SymbolicLink,
+                                    L"%wZ%wZ#%wZ%wZ",
+                                    PrefixString,
+                                    DeviceString,
+                                    GuidString,
+                                    ReferenceString);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("RtlUnicodeStringPrintf() failed with status 0x%08lx\n", Status);
+        return Status;
+    }
+
+    DPRINT("New symbolic link is '%wZ'\n", &SymbolicLink);
+
+    *SymbolicLinkName = SymbolicLink;
+    return STATUS_SUCCESS;
+}
+
+/**
+ * @brief IopSeparateSymbolicLink
+ * 
+ * Parses specified symbolic link onto the 4 parts: prefix, device string,
+ * class GUID and reference string.
+ * 
+ * @param[in] SymbolicLinkName
+ * Pointer to a symbolic link string to parse.
+ * 
+ * @param[out,opt] PrefixString
+ * Receives prefix of symbolic link. Can be '\??\' for Kernel mode or '\\?\' for User mode.
+ * 
+ * @param[out,opt] MungedString
+ * Receives Device string with '##?#' prefix at the start. Looks like ##?#ACPI#PNP0501#1#.
+ * 
+ * @param[out,opt] GuidString
+ * Receives device interface class GUID string represented by device interface.
+ * Looks like {01234567-89ab-cdef-0123-456789abcdef}.
+ * 
+ * @param[out,opt] ReferenceString
+ * Receives reference string, if any. Usually contains a human-readable
+ * subdevice name or class GUID.
+ * 
+ * @param[out,opt] ReferenceStringPresent
+ * Pointer to variable that indicates whether the reference string exists in symbolic link.
+ * TRUE if it does, FALSE otherwise.
+ * 
+ * @param[out,opt] InterfaceClassGuid
+ * Receives the interface class GUID to which specified symbolic link belongs to.
+ * 
+ * @return
+ * STATUS_SUCCESS in case of success, error NTSTATUS code otherwise.
+ *
+ */
+static
+NTSTATUS
+IopSeparateSymbolicLink(
+    _In_ PCUNICODE_STRING SymbolicLinkName,
+    _Out_opt_ PUNICODE_STRING PrefixString,
+    _Out_opt_ PUNICODE_STRING MungedString,
+    _Out_opt_ PUNICODE_STRING GuidString,
+    _Out_opt_ PUNICODE_STRING ReferenceString,
+    _Out_opt_ PBOOLEAN ReferenceStringPresent,
+    _Out_opt_ LPGUID InterfaceClassGuid)
+{
+    UNICODE_STRING KernelModePrefix = RTL_CONSTANT_STRING(L"\\??\\");
```

Comment:
These three strings can be const'ed, see the `PathSep` example above.

Reply:
Done.